### ISSUE TITLE
DOMParser parsing issues (with workaround)

### ISSUE THREAD

**Issue by chr1se** [WebReflection/linkedom#155]
I am using the DOMParser that comes in linkedom/worker.js to parse some html code, and the results are ...odd. 

```javascript

const { DOMParser:DP}  = await import("./linkedom.js"); //linkedom is in my local server, i am testing this code in browser console hence the rename to DP 
const toDom = html => (new DP()).parseFromString(html,'text/html');
const toDom2 = html => (new DOMParser()).parseFromString(html,'text/html');

{
  /******* Basic ********/
  //linkedom
  let doc = toDom(`<div></div>`); 
  doc.head;
  doc.body;
  console.log(doc.documentElement.outerHTML) //<div><head></head><body></body></div>
  console.log(doc.toString()) //<div><head></head><body></body></div>
  
  //standard
  let doc2 = toDom2(`<div></div>`); 
  doc2.head;
  doc2.body;
  console.log(doc.documentElement.outerHTML) //<html><head></head><body><div></div></body></html>
}

{
  /******* HEAD Element ********/
  //linkedom
  let doc = toDom(`<meta charset="utf8"><div></div>`); 
  doc.head;
  doc.body;
  console.log(doc.documentElement.outerHTML) //<meta charset="utf8"><head></head><body></body></meta>
  console.log(doc.toString()) //<meta charset="utf8"><head></head><body></body></meta><div></div>
  
  //standard
  let doc2 = toDom2(`<meta charset="utf8"><div></div>`); 
  doc2.head;
  doc2.body;
  console.log(doc.documentElement.outerHTML) //'<html><head><meta charset="utf8"></head><body><div></div></body></html>'

}

{
  /******* BAD HTML ********/
  //linkedom
  let doc = toDom(`<meta charset="utf8"><title></title><html><div></div></html>`); 
  doc.head;
  doc.body;
  console.log(doc.documentElement.outerHTML) //<meta charset="utf8"><head></head><body></body></meta>
  console.log(doc.toString()) //<meta charset="utf8"><head></head><body></body></meta><title></title><html><div></div></html>

  //standard
  let doc2 = toDom2(`<meta charset="utf8"><title></title><html><div></div></html>`); 
  doc2.head;
  doc2.body;
  console.log(doc.documentElement.outerHTML) //<html><head><meta charset="utf8"><title></title></head><body><div></div></body></html>
}
```
I understand that this is not meant to be totally web standards complaint but these results seem like it might just be a mistake...but i could be wrong, perhaps its intentional for performance reasons...i dunno.

I have a work-around to make it complaint, if you can improve on it please add your code below

```javascript
const  toDom3 = async function(html){
  const { DOMParser:DP}  = await import("./linkedom.js");
  const parse = (new DP()).parseFromString;
  let doc = parse(html,'text/html')

  let doc2 = parse(`<html><head></head><body></body></html>`,'text/html');
  let headElems = ['title','style','meta','link','script','base'];
  let headInsert = true;
  
  for(let e of doc.all){
    if(e.localName == 'html' || e.parentElement)continue;
    else if(headInsert && headElems.includes(e.localName))doc2.head.append(e);
    else{
      doc2.body.append(e);
      headInsert = false;
    }
  }
  return doc2;
}

```
Hope this helps

**Comment by chr1se** [comment]
crap! 
sorry my workaround is incorrect. I shouldnt be looping on ```doc.all```. 
It will insert all deeper nested children elements into the body element.

Correction coming shortly

**Comment by WebReflection** [comment]
this looks like a duplicate of https://github.com/WebReflection/linkedom/issues/152

TL;DR I have no interest, neither use cases, to parse HTML chunks. There are proper ways of doing that, and `template` nodes are just one example, everything else is abusing an API that returns `document.documentElement` that can only be an HTML node.

This is a won't fix

**Comment by chr1se** [comment]
@WebReflection  I am working on a page builder application  that serverside renders/hydrates standard web components (serververside will be in deno deploy).
When I parse a html string using that doesnt include \<head>/\<body> elements and call ```document.body.childnodes```, those elements(head & body) are injected into the dom in the wrong position ( ...they are inject not in the same place that the standard web DOMParser() would)

I spent hours trying to figure out why my app was breaking and when i spotted it and found a workaround, I thought I would help others out that may have this issue.


> There are proper ways of doing that, and template nodes are just one example

for my use case, parsing a string looking for custom elements to inject declarative shadow doms, i dont need a template node


>everything else is abusing an API that returns document.documentElement that can only be an HTML node

I dont understand how I am "abusing" the api, the examples i gave above were to demonstrate how i was generating the error, and to show how the results from linkedom DomParser differ to the web DomParser

sorry maybe i came-off rude in my first post, I appreciate your hard work creating linkedom, I wasnt asking you to fix anything, I was trying to help others that will inevitably have the same problem I had.  The issue tracker may not be the best place to put it but i dont know where-else to put my workaround.

**Comment by WebReflection** [comment]
documentElement is always an HTML node in browsers. To parse any chunk you can use template.innerHTML and the access its content fragment, you don’t need a new document or a documentElement.
